#!/usr/bin/Rscript

#==================#
#   HEADER START   #
#==================#
### Created: Dec 06, 2013
### Modified: Dec 06, 2013
### Author: Alexander Kanitz
### Company: Zavolan Group, Biozentrum, University of Basel
### Version: v1.1
### Adapted from: n/a
### Requirements: optparse
#==================#
#    HEADER END    #
#==================#


#==========================#
#   PRE-REQUISITES START   #
#==========================#

#---> LOAD LIBRARIES <---#
suppressPackageStartupMessages(library("optparse"))

#---> GET SCRIPT NAME <---#
script <- sub("--file=", "", basename(commandArgs(trailingOnly=FALSE)[4]))

#---> COMMAND-LINE ARGUMENTS <---#
## List of allowed/recognized arguments
option_list <- list(
		make_option("--overlaps", type="character", action="store", default="", help="REQUIRED: BED-like file of overlaps between regions of interest and a sequencing library; file must be of the default format generated by bedtools intersect/window with the sequencing library BED file passed as argument -a. [default: %default]", metavar="file"),
		make_option("--output", type="character", action="store", default="", help="REQUIRED: Output filename. [default: %default]", metavar="file"),
		make_option("--overlaps-minus", type="character", action="store", default="", help="If strands should be analyzed separately (options --strands, --up-down, --quadrants), a separate BED-like file is required for overlaps on the '+' and '-' strands; use --overlaps to pass only overlaps on the '+' strand in this case (format requirements: see option --overlaps). [default: %default]", metavar="file"),
		make_option("--sites", type="character", action="store", default="", help="A list of IDs/names (one per line) corresponding to the name field in the regions of interest file used to generate the file passed to --overlaps; if provided, only these sites/regions are considered for the count analysis. [default: %default]", metavar="file"),
		make_option("--controls-overlaps", type="character", action="store", default="", help="If desired, contains overlaps between the sequencing library and sets of control sites for the regions of interst (format as in option --overlaps); used for count normalization if specified. [default: %default]", metavar="file"),
		make_option("--controls-overlaps-minus", type="character", action="store", default="", help="If strands should be analyzed separately (options --strands, --up-down, --quadrants), a separate BED-like file is required for overlaps on the '+' and '-' strands; use --controls-overlaps to pass only overlaps on the '+' strand in this case (format requirements: see option --overlaps). [default: %default]", metavar="file"),
		make_option("--sites-controls-table", type="character", action="store", default="", help="A lookup table defining for each region of interest the set of control sites (format: region_of_interst TAB control_site_1|control_site_2|...|control_site_n); option is ignored if --controls-overlaps is unset; otherwise, if provided, only the control sites for your regions of interest are used. [default: %default]", metavar="file"),
		make_option("--offset", type="integer", action="store", default=NULL, help="The position relative to the middle of the regions of interest used when analyzing separately the counts that occur up-/downstream of the regions of interest (options --left-right, --up-down, --quadrants). [default: %default]", metavar="integer"),
		make_option("--crick-offset", type="integer", action="store", default=NULL, help="If strands should be analyzed separately (options --left-right, --up-down, --quadrants), allows to specify a separate offset for the minus strand (e.g. for asymmetric cutting by restriction enzymes etc.). [default: %default]", metavar="integer"),
		make_option("--reads", type="double", action="store", default=NULL, help="The total number of reads in the sequencing library; used for count normalization if specified. [default: %default]", metavar="double"),
		make_option("--alignments", type="double", action="store", default=NULL, help="The total number of alignments in the sequencing library; used for count normalization if specified. [default: %default]", metavar="double"),
		make_option("--ref-gene-names", type="character", action="store", default="", help="The names of reference genes that shall be used for normalization; names have to be separated by the pipe ('|') character; for each supplied gene name, a corresponding count has to be supplied via --ref-gene-counts. [default: %default]", metavar="string"),
		make_option("--ref-gene-counts", type="character", action="store", default="", help="Counts for each of the reference genes supplied via --ref-gene-names; separate via the pipe ('|') character. [default: %default]", metavar="string"),		
		make_option("--strands", action="store_true", default=FALSE, help="Specify this option if strand counts should be analyzed separately; requires argument to --controls-overlaps. [default: %default]"),
		make_option("--left-right", action="store_true", default=FALSE, help="Specify this option if counts originating from the 'left' and 'right' side of the regions of interest should be analyzed separately; compare with --up-down. [default: %default]"),
		make_option("--up-down", action="store_true", default=FALSE, help="Specify this option if counts originating up- and downstream of the regions of interest should be analyzed separately; compare with --left-right; requires argument to --controls-overlaps. [default: %default]"),
		make_option("--quadrants", action="store_true", default=FALSE, help="Specify this option if counts originating from the four quadrants defined by the two strands and the relative orientation to the regioin of interest should be analyzed separately; requires argument to --controls-overlaps  [default: %default]"),
		make_option("--help", action="store_true", default=FALSE, help="Show this information and die [default: %default]"),
		make_option("--usage", action="store_true", default=FALSE, dest="help", help="Show this information and die [default: %default]"),
		make_option("--verbose", action="store_false", default=TRUE, help="Print detailed status messages. [default: %default]"),
		make_option("--quiet", action="store_true", default=FALSE, dest="verbose", help="Shut up! [default: %default]")
)
## Parse command-line arguments
description <- "\nBased on the default output of 'bedtools intersect' and similar, the script analyzes the read counts around regions of interest and writes a count summary table."
opt_parser <- OptionParser(usage="Usage: %prog [OPTIONS] --overlaps [FILE] --output [FILE]", option_list = option_list, add_help_option=FALSE, description=description)
opt <- parse_args(opt_parser)

## Die if any required arguments are missing...
if 	( opt$overlaps	== "" || opt$output == "" ) { 
	write("[ERROR] Required argument(s) missing!\n\n", stderr())	
	stop(print_help(opt_parser))
}

## Check if dependencies are met
if 	( ( opt$strands || opt$`up-down` || opt$quadrants ) && opt$`overlaps-minus` == "" ) { 
	write("[ERROR] The selected option(s) require an argument to '--overlaps-minus'!\n", stderr())	
	stop(print_help(opt_parser))
}
if 	( ( opt$strands || opt$`up-down` || opt$quadrants ) && opt$`controls-overlaps` != "" && opt$`controls-overlaps-minus` == "" ) { 
	write("[ERROR] The selected option(s) require an argument to '--controls-overlaps-minus'!\n", stderr())	
	stop(print_help(opt_parser))
}
if 	( opt$`ref-gene-names` != "" && opt$`ref-gene-counts` == "" ) { 
	write("[ERROR] If an argument to '--ref-gene-names' is supplied, '--ref-gene-counts' also requires an argument!\n", stderr())	
	stop(print_help(opt_parser))
}
if 	( opt$`ref-gene-counts` != "" && opt$`ref-gene-names` == "" ) { 
	write("[ERROR] If an argument to '--ref-gene-counts' is supplied, '--ref-gene-names' also requires an argument!\n", stderr())	
	stop(print_help(opt_parser))
}

## Check whether the supplied reference gene names and counts are compatible
if ( opt$`ref-gene-names` != "" ) {
	ref_cts <- type.convert(unlist(strsplit(opt$`ref-gene-counts`, "|", fixed=TRUE)), as.is=TRUE)
	ref_nms <- unlist(strsplit(opt$`ref-gene-names`, "|", fixed=TRUE))
	if ( ! is.numeric(ref_cts) ) {
		write("[ERROR] Cannot properly parse argument to '--ref-gene-counts'! Only use numeric values, separated by the pipe ('|') character.\n", stderr())	
		stop(print_help(opt_parser))		
	}
	if ( length(ref_cts) != length(ref_nms) ) {
		write("[ERROR] Different number of elements in the arguments to '--ref-gene-counts' and '--ref-gene-names'! Supply exactly one count per gene.\n", stderr())	
		stop(print_help(opt_parser))		
	}
}

## Non-fatal sanity checks
if 	( opt$`sites-controls-table` != "" && opt$`controls-overlaps` == "" ) { 
	write("[WARNING] Option '--controls-overlaps' not selected. Option '--sites-controls-table' is ignored!", stderr())	
}
if 	( ! ( opt$`left-right` || opt$`up-down` || opt$quadrants ) && ! is.null(opt$offset) ) { 
	write("[WARNING] Required analysis does not depend on the exact cut site position. The specified custom value for option '--offset' is ignored!", stderr())	
}
if 	( ! ( opt$strands || opt$`left-right` || opt$`up-down` || opt$quadrants ) && ! is.null(opt$`crick-offset`) ) { 
	write("[WARNING] Required analysis does not depend on the exact cut site position. The specified custom value for option '--crick-offset' is ignored!", stderr())	
}

## Set offsets to 0 if not user-specified
if 	( ( opt$`left-right` || opt$`up-down` || opt$quadrants ) && is.null(opt$offset) ) { 
	opt$offset <- 0
}
if 	( ( opt$`left-right` || opt$`up-down` || opt$quadrants ) && is.null(opt$`crick-offset`) ) { 
	opt$`crick-offset` <- opt$offset
}

#==========================#
#    PRE-REQUISITES END    #
#==========================#


#================#
#   MAIN START   #
#================#

#---> START MESSAGE <---#
if ( opt$verbose ) cat("Starting '", script, "'...\n", sep="")

#---> LOAD OVERLAPS <---#
# Print status message
if ( opt$verbose ) cat("Reading file '", basename(opt$overlaps), "'...\n", sep="")
# Read file
overlaps_sites <- read.delim(opt$overlaps, header=FALSE, stringsAsFactors=FALSE)

#---> LOAD OVERLAPS FOR MINUS STRAND IF APPLICABLE <---#
if 	( opt$`overlaps-minus` != "" ) {
	# Print status message
	if ( opt$verbose ) cat("Reading file '", basename(opt$`overlaps-minus`), "'...\n", sep="")
	# Read file
	overlaps_sites_minus <- read.delim(opt$`overlaps-minus`, header=FALSE, stringsAsFactors=FALSE)
}

#---> LOAD CONTROL SITE OVERLAPS IF APPLICABLE <---#
if 	( opt$`controls-overlaps` != "" ) {
	# Print status message
	if ( opt$verbose ) cat("Reading file '", basename(opt$`controls-overlaps`), "'...\n", sep="")
	# Read file
	overlaps_ctrls <- read.delim(opt$`controls-overlaps`, header=FALSE, stringsAsFactors=FALSE)
}

#---> LOAD CONTROL SITE OVERLAPS FOR MINUS STRAND IF APPLICABLE <---#
if 	( opt$`controls-overlaps-minus` != "" ) {
	# Print status message
	if ( opt$verbose ) cat("Reading file '", basename(opt$`controls-overlaps-minus`), "'...\n", sep="")
	# Read file
	overlaps_ctrls_minus <- read.delim(opt$`controls-overlaps-minus`, header=FALSE, stringsAsFactors=FALSE)
}

#---> GET SITES OF INTEREST IF APPLICABLE <---#
if 	( opt$sites != "" ) {
	# Print status message
	if ( opt$verbose ) cat("Reading file '", basename(opt$sites), "'...\n", sep="")
	# Read file
	sites <- unlist(read.delim(opt$sites, header=FALSE, stringsAsFactors=FALSE), use.names=FALSE)
}

#---> GET APPROPRIATE CONTROL SITES IF APPLICABLE <---#
if 	( opt$`sites-controls-table` != "" ) {
	# Print status message
	if ( opt$verbose ) cat("Reading file '", basename(opt$`sites-controls-table`), "'...\n", sep="")
	# Read file
	sites_ctrls_tab <- read.delim(opt$`sites-controls-table`, header=FALSE, stringsAsFactors=FALSE)
	# Extract appropriate control sites
	ctrls <- unlist(strsplit((sites_ctrls_tab[sites_ctrls_tab[,1] %in% sites, 2]), "\\|"))
}

#---> SUBSET SITES OF INTEREST IF APPLICABLE <---#
if 	( opt$sites != "" ) {
	# Print status message
	if ( opt$verbose ) cat("Subsetting sites...\n", sep="")
	# Subset sites of interest
	overlaps_sites <- overlaps_sites[overlaps_sites[,11] %in% sites,]
	if ( opt$`overlaps-minus` != "" ) overlaps_sites_minus <- overlaps_sites_minus[overlaps_sites_minus[,11] %in% sites,]
}

#---> SUBSET CONTROL SITES OF INTEREST IF APPLICABLE <---#
if 	( opt$`sites-controls-table` != "" ) {
	# Print status message
	if ( opt$verbose ) cat("Subsetting control sites...\n", sep="")
	# Subset control sites of interest
	overlaps_ctrls <- overlaps_ctrls[overlaps_ctrls[,11] %in% ctrls,]
	if ( opt$`controls-overlaps-minus` != "" ) overlaps_ctrls_minus <- overlaps_ctrls_minus[overlaps_ctrls_minus[,11] %in% ctrls,]
}

#---> CALCULATE DISTANCES FROM REGIONS OF INTEREST (CENTER +/- OFFSET) IF APPLICABLE <---#
if 	( opt$`left-right` || opt$`up-down` || opt$quadrants ) {
	# Print status message
	if ( opt$verbose ) cat("Calculating distances of reads/tags from regions of interest...\n", sep="")
	# Calculate distances
	dist <- ((overlaps_sites[,3] + overlaps_sites[,2]) / 2) - ((overlaps_sites[,10] + overlaps_sites[,9]) / 2 + opt$offset)
	if ( opt$`overlaps-minus` != "" ) dist_minus <- ((overlaps_sites_minus[,3] + overlaps_sites_minus[,2]) / 2) - ((overlaps_sites_minus[,10] + overlaps_sites_minus[,9]) / 2 + opt$`crick-offset`)
}

#---> CALCULATE CONTROL SITE COUNTS <---#
if 	( opt$`controls-overlaps` != "" ) {
	# Print status message
	if ( opt$verbose ) cat("Calculating control site counts...\n", sep="")
	# Calculate distances
	ctrl_site_counts <- nrow(overlaps_ctrls)
	if ( opt$`controls-overlaps-minus` != "" ) ctrl_site_counts <- ctrl_site_counts + nrow(overlaps_ctrls_minus)
}

#---> CALCULATE TOTAL COUNTS <---#
# Print status message
if ( opt$verbose ) cat("Calculating counts...\n", sep="")
# Calculate counts
t <- nrow(overlaps_sites)
if ( opt$strands || opt$`up-down` || opt$quadrants ) t <- t + nrow(overlaps_sites_minus)
groups <- "TOTAL"
counts <- t

#---> CALCULATE STRAND COUNTS <---#
if ( opt$strands ) {
	# Print status message
	if ( opt$verbose ) cat("Calculating +/- strand counts...\n", sep="")
	# Calculate counts
	p <- nrow(overlaps_sites)
	m <- nrow(overlaps_sites_minus)
	groups <- c(groups, "Plus", "Minus")
	counts <- c(counts, p, m)
}

#---> CALCULATE LEFT/RIGHT COUNTS <---#
if ( opt$`left-right` ) {
	# Print status message
	if ( opt$verbose ) cat("Calculating left/right counts...\n", sep="")
	# Calculate counts
	l <- length(dist[(dist < 0)])
	if ( opt$`overlaps-minus` != "" ) l <- l + length(dist_minus[(dist_minus < 0)])
	r <- length(dist[(dist > 0)])
	if ( opt$`overlaps-minus` != "" ) r <- r + length(dist_minus[(dist_minus > 0)])
	groups <- c(groups, "Left", "Right")
	counts <- c(counts, l, r)
}

#---> CALCULATE UP-/DOWNSTREAM COUNTS <---#
if ( opt$`up-down` ) {
	# Print status message
	if ( opt$verbose ) cat("Calculating up-/downstream counts...\n", sep="")
	# Calculate counts
	u <- length(dist[(dist < 0)]) + length(dist_minus[(dist_minus > 0)])
	d <- length(dist[(dist > 0)]) + length(dist_minus[(dist_minus < 0)])
	groups <- c(groups, "Upstream", "Downstream")
	counts <- c(counts, u, d)
}

#---> CALCULATE QUADRANT COUNTS <---#
if ( opt$quadrants ) {
	# Print status message
	if ( opt$verbose ) cat("Calculating quadrant counts...\n", sep="")
	# Calculate counts
	q1 <- length(dist[(dist > 0)])
	q2 <- length(dist_minus[(dist_minus > 0)])
	q3 <- length(dist_minus[(dist_minus < 0)])
	q4 <- length(dist[(dist < 0)])
	groups <- c(groups, "Quadrant 1", "Quadrant 2", "Quadrant 3", "Quadrant 4")
	counts <- c(counts, q1, q2, q3, q4)
}

#---> BUILD DATA FRAME <---#
# Print satatus message
if ( opt$verbose ) cat("Building output table...\n", sep="")
# Generate data frame
count_summary <- data.frame(Counts=counts)
# Set rownames
rownames(count_summary) <- groups
# If applicable, normalize counts by total read counts and add to data frame
if ( opt$reads != "" ) { 
	CPM_reads <- counts / opt$reads * 1000000
	count_summary <- cbind(count_summary, CPM_reads)
}
# If applicable, normalize counts by total alignment counts and add to data frame
if ( opt$alignments != "" ) {
	CPM_alignments <- counts / opt$alignments * 1000000
	count_summary <- cbind(count_summary, CPM_alignments)
}
# If applicable, normalize counts by control site counts and add to data frame
if ( opt$`controls-overlaps` != "" ) {
	Norm_by_CTRL_site_counts <- counts / ctrl_site_counts
	count_summary <- cbind(count_summary, Norm_by_CTRL_site_counts)
}
# If applicable, normalize counts by reference gene counts and add to data frame
if ( opt$`ref-gene-names` != "" ) {
	ref_counts <- list()
	for (gene in 1:length(ref_nms)) {
		ref_counts[[ref_nms[gene]]] <- counts / ref_cts[gene]
	}
	count_summary <- cbind(count_summary, lapply(ref_counts, "["))
}

#---> EXPORT OUTPUT <---#
# Print status message
if ( opt$verbose ) cat("Writing results to file '", opt$output, "'...\n", sep="")
# Use rtracklayer::export method to export GRanges object as BED file
write.table(count_summary, opt$output, quote=FALSE, sep="\t")

#---> END MESSAGE <---#
if ( opt$verbose ) cat("Done.\n")

#================#
#    MAIN END    #
#================#