#######
### GENERAL:
### --------
### Author: 		Alexander Kanitz
### Created: 		05-APR-2013
### Modified:		05-APR-2013
### Language: 		R
### Version:		2.15.2
### Requirements:	Bioconductor_2.11, edgeR + dependencies
### Description: 	Compute relative log2 fold changes between two edgeR analyses (query and subject); query feature names are tested for suffixes added by base::make.unique ([name].[serial_number]) before looking up corresponding values in subject to allow for multiple queries matching to a single subject (e.g. exon to gene comparison) 
### Arguments: 		1. Query file; 2. Subject file; 3. Output file prefix (may contain path!); 1./2. are tables of differentially expressed features generated by "edgeR_analysis.R" (column 1 = features [rownames], column 2 = log2 FC); possible duplicate query feature names are allowed in principal, but have to have beeen made unique with base::make.unique BEFORE using this script!
### Output: 		Table with relative log2 fold changes for each query feature 
### Usage example:	Rscript compare_edgeR.R ./query.tab ./subject.tab path/to/output/files/prefix
#######

### A. Pre-requisites
# Get command line arguments
args <- commandArgs(trailingOnly=TRUE)
## Pass arguments
query <- args[1]
subject <- args[2]
prefix <- args[3]
###

### B. Load data
# Load query data frame
query_df <- read.table(query)
# Load subject data frame
subject_df <- read.table(subject)
###

### C. Generate table of relative fold changes
# Create (ambiguous) query feature names from unique identifiers
feature_names <- sub('\\.\\d+$', "", rownames(query_df))
# Create vector of subject indices for each query feature name
lookup_indices <- match(feature_names, rownames(subject_df))
# Calculate relative log2 fold changes between query and subject (differences of log2 query and log2 subject)
rel_fc_query <- query_df[,1] - subject_df[lookup_indices,1]
# Bind (ambiguous) identifiers and relative log2 fold changes to data.frame
rel_fc_query_df <- data.frame("Features"=feature_names, "Log2FC"=rel_fc_query, row.names=row.names(query_df), stringsAsFactors=FALSE, check.rows=TRUE)
###

### D. Write output table and plot results
# Write table of relative fold changes
write.table(rel_fc_query_df, file=paste(prefix, "relative_fold_changes.tsv", sep="_"), quote=FALSE, sep="\t")
# Write summary to <STDOUT>
quantile(rel_fc_query_df[,2], c(.01, .02, .05, .1, .15, .25, .5, .75, .85, .90, .95, .98, .99))
# Plot histogram
pdf(file=paste(prefix, "relative_fold_changes.pdf", sep="_"), width = 6, height = 6)
breaks <- seq(from=floor(min(rel_fc_query_df[,2])), to=ceiling(max(rel_fc_query_df[,2])), by=1)
hist(rel_fc_query_df[,2], breaks=breaks, freq=FALSE, xlab="Relative log2 fold change")
dev.off()
###

### E. Write output & save session image
# Remove unused/temp variables
rm(feature_names, lookup_indices, rel_fc_query, breaks)
# Save workspace image/session
save.image(file=paste(prefix, "image.Rdata", sep="_"))
###
